test/Microsoft.AspNet.Http.Tests/DefaultHttpContextTests.cs

```diff
@@ -149,93 +149,88 @@ public void SetItems_NewCollectionUsed()
+            context.Initialize(newFeatures);
+            TestAllCachedFeaturesAreNull(context, newFeatures);
+            Assert.Equal(3, newFeatures.Count());
 
-            Assert.NotEqual(0, newFeatures.Count());
+            // getting feature properties populates new feature collection with defaults
+            TestAllCachedFeaturesAreSet(context, newFeatures);
+            Assert.NotEqual(3, newFeatures.Count());
         }
 
-        void TestCachedFeaturesAreNull(Type type, object value, IFeatureCollection features)
+        void TestAllCachedFeaturesAreNull(HttpContext context, IFeatureCollection features)
         {
+            TestCachedFeaturesAreNull(context, features);
+            TestCachedFeaturesAreNull(context.Request, features);
+            TestCachedFeaturesAreNull(context.Response, features);
+            TestCachedFeaturesAreNull(context.Authentication, features);
+            TestCachedFeaturesAreNull(context.Connection, features);
+            TestCachedFeaturesAreNull(context.WebSockets, features);
+        }
 
-            var fields = type
+        void TestCachedFeaturesAreNull(object value, IFeatureCollection features)
+        {
+            var type = value.GetType();
+
+            var field = type
                 .GetFields(BindingFlags.NonPublic | BindingFlags.Instance)
-                .Where(f => f.FieldType.GetTypeInfo().IsInterface);
+                .Single(f => 
+                    f.FieldType.GetTypeInfo().IsGenericType &&
+                    f.FieldType.GetGenericTypeDefinition() == typeof(FeatureReferences<>));
 
-            foreach (var field in fields)
-            {
-                if (field.FieldType == typeof(IFeatureCollection))
-                {
-                    Assert.Same(features, field.GetValue(value));
-                }
-                else
-                {
-                    Assert.Null(field.GetValue(value));
-                }
-            }
+            var boxedExpectedStruct = features == null ?
+                Activator.CreateInstance(field.FieldType) :
+                Activator.CreateInstance(field.FieldType, features);
+
+            var boxedActualStruct = field.GetValue(value);
+
+            Assert.Equal(boxedExpectedStruct, boxedActualStruct);
         }
 
-        void TestCachedFeaturesAreSet(Type type, object value, IFeatureCollection features)
+        void TestAllCachedFeaturesAreSet(HttpContext context, IFeatureCollection features)
         {
+            TestCachedFeaturesAreSet(context, features);
+            TestCachedFeaturesAreSet(context.Request, features);
+            TestCachedFeaturesAreSet(context.Response, features);
+            TestCachedFeaturesAreSet(context.Authentication, features);
+            TestCachedFeaturesAreSet(context.Connection, features);
+            TestCachedFeaturesAreSet(context.WebSockets, features);
+        }
+
+        void TestCachedFeaturesAreSet(object value, IFeatureCollection features)
```

Comment:
I was trying to test if the cached (private var) was correct for the current and any future additions so act as a regression safety test; even found a few instances where it wasn't actually being saved to the private var with the test.